lib/Runtime/Library/ArrayBuffer.cpp

```diff
@@ -1042,68 +1047,89 @@ namespace Js
         }
         uint32 growSize = newBufferLength - this->bufferLength;
 
-        bool failedReport = false;
-        const auto reportFailedFn = [&failedReport] { failedReport = true; };
+        struct AutoFailFastOnBadState
+        {
+            bool isStateValid = true;
+            ~AutoFailFastOnBadState()
+            {
+                if (!isStateValid)
+                {
+                    // We ended up in an unrecoverable state, fail fast
+                    Js::Throw::FatalInternalError();
+                }
+            }
+        };
+        AutoFailFastOnBadState autoFailFastOnBadState;
 
         WebAssemblyArrayBuffer* newArrayBuffer = nullptr;
 #if ENABLE_FAST_ARRAYBUFFER
         if (CONFIG_FLAG(WasmFastArray))
         {
             AssertOrFailFast(this->buffer);
-            ReportDifferentialAllocation(newBufferLength, reportFailedFn);
-            if (failedReport)
-            {
-                return nullptr;
-            }
-
             if (growSize > 0)
             {
-                LPVOID newMem = VirtualAlloc(this->buffer + this->bufferLength, growSize, MEM_COMMIT, PAGE_READWRITE);
-                if (!newMem)
+                const auto virtualAllocFunc = [&]
+                {
+                    return !!VirtualAlloc(this->buffer + this->bufferLength, growSize, MEM_COMMIT, PAGE_READWRITE);
+                };
+                if (!this->GetRecycler()->DoExternalAllocation(growSize, virtualAllocFunc))
                 {
-                    Recycler* recycler = this->GetRecycler();
-                    recycler->ReportExternalMemoryFailure(newBufferLength);
                     return nullptr;
                 }
             }
-            newArrayBuffer = GetLibrary()->CreateWebAssemblyArrayBuffer(this->buffer, newBufferLength);
+            newArrayBuffer = this->GetLibrary()->CreateWebAssemblyArrayBuffer(this->buffer, newBufferLength);
         }
         else
 #endif
         if (this->GetByteLength() == 0)
         {
             if (growSize > 0)
             {
-                newArrayBuffer = GetLibrary()->CreateWebAssemblyArrayBuffer(newBufferLength);
+                // Creating a new buffer will do the external memory allocation report
+                newArrayBuffer = this->GetLibrary()->CreateWebAssemblyArrayBuffer(newBufferLength);
             }
             else
             {
-                newArrayBuffer = GetLibrary()->CreateWebAssemblyArrayBuffer(this->buffer, 0);
+                newArrayBuffer = this->GetLibrary()->CreateWebAssemblyArrayBuffer(this->buffer, 0);
             }
         }
         else
         {
-            ReportDifferentialAllocation(newBufferLength, reportFailedFn);
-            if (failedReport)
+            // Disable QC while doing a ReAlloc to minimize chances to end up in a bad state
+            AutoDisableInterrupt autoDisableInterrupt(this->GetScriptContext()->GetThreadContext()->GetInterruptPoller(), true);
```

Comment:
>true [](start = 122, length = 4)

I think you need to pass false here - as you don't want explicit completion. True here means that you are going to call the autoDisableInterrupt.Completed() in this block.